CONTRIBUTING.md

```diff
@@ -66,6 +66,111 @@ For first-time contributers, we suggest to look for issues categorized as [trivi
 
 If you want a hassle-free contribution look for issues categorized as [preapproved](https://issues.dlang.org/buglist.cgi?component=dmd&keywords=preapproved&product=D).
 
+## DMD Style
+
+Here is a short list of stylistic issues the core team will expect in
+pull requests. Much of the source code does not follow these, but
+we expect new code to, and PRs retrofitting the existing code to
+follow it is welcome.
+
+1. Use attributes `const`/`nothrow`/`pure`/`scope`/`@safe`/`private`/etc.
+Successfully using `pure` functions is regarded with particular favor.
+
+2. Use correct Ddoc function comment blocks. Do not use Ddoc comments for
+overrides unless the overriding function does something different (as far as
+the caller is concerned) than the overridden function. Ddoc comment blocks
+are often overkill for nested functions and function literals; use ordinary
+comments for those.
+
+3. Do not use `strlen`/`strcmp` and their ilk. Use D arrays instead.
+
+4. Use `ref` instead of raw pointers.
+
+5. Use nested functions to get rid of rats' nests of goto's.
+
+6. Look for duplicative code and factor out into functions.
+
+7. Declare local variables as `const` as much as possible.
+
+8. Use Single Assignment for local variables:
+```
+T t = x;
+...
+t = y;
+ ...
+```
+becomes:
+```
+T tx = x;
+...
+T ty = y;
+...
+```
+
+9. "Shrinkwrap" the scope of local variables as tightly as possible
+around their uses.
+
+10. Similar to (8), use distinct variable names for non-overlapping uses.
+
+11. Avoid use of mutable globals as much as practical. Consider passing them
+in as parameters.
+
+12. Avoid use of default parameters. Spell them out.
+
+13. Minimize use of overloading.
+
+14. Avoid clever code. Anybody can write clever code. It takes genius to write
+simple code.
+
+15. Try to reduce cyclomatic complexity, i.e. think about how to make the code work
+without control flow statements.
+
+16. Try not to mix functions that "answer a question" with functions that
+"mutate the data".
+This was done successfully in src/dmd/escape.d, it wasn't easy, but
+it was well worth it.
+
+17. Try to eliminate reliance on `global.errors`.
```

Comment:
Can we change this to `global`, or is it too drastic?

Reply:
There's a particular problem with `global.errors`